### ISSUE TITLE
Registration Process Flow

### ISSUE THREAD

**Issue by bramus** [ocpi/ocpi#166]
Hi,

Upon reading the OCPI spec I started pondering on the Registration Process Flow described in it. 

But let me detour first: Say I want to sign up for some service – outside of the OCPI spec, just think of any service – which uses invite codes. 

What happens most of the time is that the party who offers the service (e.g. the _DataProvider_) sends an access token (~ invite code) to the user (e.g. the _DataConsumer_). The Consumer can then register himself using said invite code. After registration the DataConsumer can then access the services offered by the DataProvider.

Put differently: it's the user of the the services provided that initialises the registration process.

---

Now, linking back to the OCPI and the [Registration Process Flow](https://github.com/ocpi/ocpi/blob/master/credentials.md#31-registration). The current version (OCPI 2.1.1) reads:

> To register a CPO in an eMSP platform (or vice versa), the CPO must create a unique token that can be used for authenticating the eMSP. This token along with the versions endpoint should be sent to the eMSP in a secure way that is outside the scope of this protocol.

In this setup this – to me, correct me if I'm wrong – translates to the CPO being the DataConsumer and the eMSP the DataProvider _(For example: the CPO that will access the Tokens module provided by the eMSP)_ … **yet it's the DataConsumer which is generating the ”invite code“ (`TOKEN_A` from the spec)**.

As per spec the flow [as pictured in this lovely chart](https://github.com/ocpi/ocpi/blob/master/data/registration-sequence.png) then continues, in three main parts _(to be clear: DataProvider = eMSP, DataConsumer = CPO)_:
1. The DataProvider queries the DataConsumer (using `TOKEN_A`) to find a matching OCPI version.
2. The DataProvider registers itself by doing a `POST` to the Credentials Module of the DataConsumer. Along with that it sends `TOKEN_B` for the DataConsumer to use for returning communication.
3. The DataConsumer queries the DataProvider (using `TOKEN_B`) to find a matching version and finally returns `TOKEN_C` for the DataProvider to use in future communication (thus deprecating `TOKEN_A`.

---

The flow described above, as interpreted from the spec, feels like a reversed process to me. 

Linking back to the detour at the beginning it seems more obvious to let the DataProvider generate an access token (e.g. `TOKEN_A`) for the DataConsumer to use when registering itself with the DataProvider.

In words _(rewriting the part of the spec quoted above)_ that would read:
> To register a CPO in an eMSP platform (or vice versa), the **eMSP** must create a unique token that can be used for **registering** the **CPO**. This token along with the versions endpoint should be sent to the **CPO** in a secure way that is outside the scope of this protocol.

In that case the flow (following the same flowchart) seems way more logic _(to be clear: DataProvider = eMSP, DataConsumer = CPO)_:
1. The DataConsumer queries the DataProvider (using `TOKEN_A`) to find a matching OCPI version.
2. The DataConsumer registers itself by doing a `POST` to the Credentials Module of the DataProvider. Along with that it sends `TOKEN_B` for the DataProvider to use for returning communication.
3. The DataProvider queries the DataConsumer (using `TOKEN_B`) to find a matching version and finally returns `TOKEN_C` for the DataConsumer to use in future communication (thus deprecating `TOKEN_A`.

---

I know, the spec reads this too:

> Due to its symmetric nature, the CPO and eMSP can be swapped in the registration sequence.

This however would only change the positions of the CPO and eMPS in [the flowchart](https://github.com/ocpi/ocpi/blob/master/data/registration-sequence.png), but wouldn't change their DataConsumer/DataProvider roles.

---

Looking forward to hearing input from other people on this. Perhaps I'm overlooking something? Of perhaps there's a very good reason behind this reversed-flow that I'm not aware of?

Regards,
Bramus.

**Comment by rolandschroth** [comment]
Hi,
as CPO and EMSP are always both DataProvider and DataConsumer, therefore it doesn't really matter. Per specification, both also have to implement the Credentials endpoint to the full extent.
So for two specific partners, willing to communicate via OICP, all that remains for the registration process to work, is an agreement about whom of the two will create the initial `TOKEN_A`.

As in contrast to a publicly available service with self-registration, OCPI (from my understanding) always requires both parties to agree upon a relationship before initiating the technical registration process.
So from my point of view, there is no "reverse-flow".

I think, you partly misinterpreted the sequence-diagram.

> 2. The DataConsumer registers itself by doing a `POST` to the Credentials Module of the DataProvider. Along with that it sends `TOKEN_B` for the DataProvider to use for returning communication.
> 3. The DataProvider queries the DataConsumer (using `TOKEN_B`) to find a matching version and finally returns `TOKEN_C` for the DataConsumer to use in future communication (thus deprecating `TOKEN_A`.

The return of `TOKEN_C` (point 3) is the actual response to the `POST` to credentials module (point 2). The intermediate GET requests are only for gathering all data that is required to be stored together with `TOKEN_B`. 
So `TOKEN_A` is only used for initialization of the registration and `TOKEN_C` is provided in direct exchange for `TOKEN_B`.

Did that clarify the process? Or did I not get the point here?

Kind regards, Roland

**Comment by f7n** [comment]
Hi @bramus, is the registration flow any more clearer for you? I'm also getting similarly confused about the flow (and the Tokens, mentioned in your issue #164). Mainly I'm confused about the exchange of tokens. The image provided isn't very clear to me

As per the documentation for the GET method:

>Retrieves the credentials object to access the server's platform. *The request body is empty*, the response *contains the credentials object to access the server's platform*. This credentials object also contains extra information about the server such as its business details.

So a GET can be performed on the `/credentials` endpoint, and the server should respond with `TOKEN_C`? Or the server should respond with only business details and a URL to the server's OCPI versions? Or should the server respond with `TOKEN_C`? This doesn't seem secure, but the documentation says that " the response contains the credentials object to access the server's platform".

Moving on to the POST method (client->server), let me know if I understand this correctly. `TOKEN_B` is sent in a POST request to the server from the the client (along with the client's business details), and the server will respond with `TOKEN_C` along with the server's business details. Is `TOKEN_B` then discarded by the server? Or should it be stored for use in the future?

So what's the point in the GET method? That's my main confusion. How should the server respond to a `/credentials` GET request?

So very confused.

**Comment by bramus** [comment]
@f7n The `TOKEN_B` that's being sent with the `POST` request should be stored by the party that's handling the request. They need it for further use, when communicating back to the originator.

Take the `Locations` module for example: A CPO can send over it's locations to an MSP, but an MSP can also request a location at a CPO. They each have their own token token to use (`TOKEN_B` and `TOKEN_C`).

**Comment by Mathieu-Castets** [comment]
No, Credential object must contain a token. It should return the same token the party used in the Authorization header.

**Comment by f7n** [comment]
Okay, fine. And the Authorization header token is different from TOKEN_A/TOKEN_B/TOKEN_C right?  

(https://github.com/ocpi/ocpi/blob/master/transport_and_format.md#request-format)[Here it says each request must include a Authorization header.] But I'm unsure if this should change as TOKEN_A changes to B, C etc.

**Comment by Mathieu-Castets** [comment]
No, it's not a different thing. Authorization header carries the token given by the other party. 

When a query hits your server, you always look at the token in the Authorization header and search in your database which operator match to this token. Thus, you can identify the client and authorize its action or you may return a 401 Unauthorized.

**Comment by KevBelisle** [comment]
We have recently been working on implementing the standard and this is my understanding of the registration process between companies Gamma and Muon :

1. Gamma generates Token_A and gives it to Muon offline (email, snail mail, whatever) along with their (Gamma's) versions endpoint.

2. Muon sends a GET request to Gamma/ocpi/cpo/versions with Token_A in the Authorization header. Gamma validates the token in the Auth header, knows that the request comes from Muon and responds with available versions.

3. Muon chooses the most recent compatible version - in this case 2.1.1.

3. Muon sends a GET request to Gamma/ocpi/cpo/2.1.1/ with Token_A in the Authorization header. Gamma validates the token in the Auth header, knows that the request comes from Muon and responds with the details for version 2.1.1.

4. Muon stores the version and endpoints he just received from Gamma.

5. Muon generates Token_B for Gamma.

6. Muon makes a POST request to Gamma/ocpi/cpo/2.1.1/credentials with Token_A in the Authorization header, and with a JSON request body that includes Token_B and their (Muon's) version endpoint.

**Before responding to this POST request, Gamma does a lot of things...**

7. Gamma sends a GET request to Muon/ocpi/cpo/versions with Token_B in the Authorization header. Muon validates the token in the Auth header, knows that the request comes from Gamma and responds with available versions.

8. Gamma knows to use version 2.1.1 since the currently pending POST request from Muon (initiated in Step 6) was on their 2.1.1 endpoint.

9. Gamma sends a GET request to Muon/ocpi/cpo/2.1.1/ with Token_B in the Authorization header. Muon validates the token in the Auth header, knows that the request comes from Gamma and responds with the details for version 2.1.1.

10. Gamma stores the version and endpoints they just received from Muon.

11. Gamma stores Token_B for use with all further requests to Muon.

12. Gamma generates Token_C that will replace Token_A that was initially exchanged offline.

**Gamma is now ready to respond to Muon's POST request initiated in step 6!**

6.2. Gamma responds to Muon's request with Token_C.

13. Muon stores Token_C and discards Token_A.

14. All future requests from Muon to Gamma will now be Authorized with Token_C, and all future requests from Gamma to Muon will be Authorized with Token_B.
Token_A is discarded and not to be used again.



I did not find the sequence diagram in the documentation particularly clear. I think the attached image makes the sequence steps 6-12 clearer.

![ocpi_registration_sequence](https://user-images.githubusercontent.com/5402931/40204069-498228a0-59f5-11e8-890a-577676042255.jpg)





Hopefully that helps someone understand the sequence.

**Comment by KevBelisle** [comment]
When I say "sends a request with Token_A in the Authorization header" I mean the following :

If we suppose that Token_A is the string "ABCDEFGHIJKLMNOP", the HTTP request would include the following header :

Authorization: Token ABCDEFGHIJKLMNOP

If you were testing this in Postman it would look like this:
![image](https://user-images.githubusercontent.com/5402931/40204238-d1394d64-59f5-11e8-8ca2-a4fcb0018950.png)

**Comment by RobertdeLeeuw-IHomer** [comment]
See also #164
